modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DataNodesManager.java

```diff
@@ -0,0 +1,1562 @@
+                        }
+                    })
+                    .thenCompose(ignored -> nullCompletedFuture());
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    private void stopAllTimers(int zoneId) {
+        ZoneTimers zt = zoneTimers.get(zoneId);
+        if (zt != null) {
+            zt.stopAllTimers();
+        }
+    }
+
+    private WatchListener createScaleUpTimerPrefixListener() {
+        return event -> inBusyLockAsync(
+                busyLock,
+                () -> processWatchEvent(
+                        event,
+                        DISTRIBUTION_ZONE_SCALE_UP_TIMER_PREFIX,
+                        (zoneId, e) -> processTimerWatchEvent(zoneId, e, true)
+                )
+        );
+    }
+
+    private WatchListener createScaleDownTimerPrefixListener() {
+        return event -> inBusyLockAsync(
+                busyLock,
+                () -> processWatchEvent(
+                        event,
+                        DISTRIBUTION_ZONE_SCALE_DOWN_TIMER_PREFIX,
+                        (zoneId, e) -> processTimerWatchEvent(zoneId, e, false)
+                )
+        );
+    }
+
+    private WatchListener createDataNodesListener() {
+        return event -> inBusyLockAsync(
+                busyLock,
+                () -> processWatchEvent(event, DISTRIBUTION_ZONE_DATA_NODES_HISTORY_PREFIX, this::processDataNodesHistoryWatchEvent)
+        );
+    }
+
+    private CompletableFuture<Void> processTimerWatchEvent(int zoneId, Entry e, boolean scaleUp) {
+        if (e.tombstone()) {
+            return nullCompletedFuture();
+        }
+
+        DistributionZoneTimer timer = DistributionZoneTimerSerializer.deserialize(e.value());
+
+        CatalogZoneDescriptor zoneDescriptor = catalogManager.activeCatalog(e.timestamp().longValue()).zone(zoneId);
+
+        if (zoneDescriptor == null) {
+            return nullCompletedFuture();
+        }
+
+        if (scaleUp) {
+            onScaleUpTimerChange(zoneDescriptor, timer);
+        } else {
+            onScaleDownTimerChange(zoneDescriptor, timer);
+        }
+
+        return nullCompletedFuture();
+    }
+
+    private CompletableFuture<Void> processDataNodesHistoryWatchEvent(int zoneId, Entry e) {
+        if (!e.tombstone() && e.value() != null) {
+            DataNodesHistory history = DataNodesHistorySerializer.deserialize(e.value());
+
+            CatalogZoneDescriptor zoneDescriptor = catalogManager.activeCatalog(e.timestamp().longValue()).zone(zoneId);
+
+            if (zoneDescriptor == null) {
+                return nullCompletedFuture();
+            }
+
+            dataNodesHistoryVolatile.put(zoneId, history);
+        } else if (e.tombstone()) {
+            dataNodesHistoryVolatile.remove(zoneId);
+        }
+
+        return nullCompletedFuture();
+    }
+
+    private static CompletableFuture<Void> processWatchEvent(
+            WatchEvent event,
+            String keyPrefix,
```

Comment:
You can pass `byte[]` here instead, it will save us some unnecessary string conversions

Reply:
fixed